libunwind/src/UnwindRegistersSave.S

```diff
@@ -885,7 +898,7 @@ DEFINE_LIBUNWIND_FUNCTION(__unw_getcontext)
   str    d30,     [c0, #0x0f0]
   str    d31,     [c0, #0x0f8]
   mov    x0, #0                   // return UNW_ESUCCESS
-  ret
+  b      _rtld_unw_getcontext
```

Comment:
Thinking about this more, won't PLTs and lazy binding screw this up? c16 and c17 can be clobbered by PLTs, and lazy binding can clobber more.

Reply:
I'm not sure I follow. `_rtld_unw_getcontext` doesn't use c16, and `_rtld_unw_setcontext` only uses it inside the function itself to temporarily store the value of c2. Why does it matter if the PLT makes use of c16 and c17?

In the `unw_getcontext` path we `stp c16,c17, [c0, #0x100]` first, then we branch to `_rtld_unw_getcontext`, going through the PLT (which clobbers c16 and c17, but we have the old values at c0+0x100), but gives us a sealed executive stack pointer.

In `_rtld_unw_setcontext`, c16 and c17 don't get restored anyway since they were clobbered (there's already a comment stating that there), then we go through the PLT and clobber them again, and then in the assembly inside rtld we clobber c16 yet again to temporarily store the value of c2.

Perhaps I'm missing something here? If you mean that the call to `unw_getcontext` would go via the PLT, that is handled by marking `unw_getcontext` as a trusted function in the c18n policy, so it should never go through the PLT.

Reply:
Oh I meant to comment on the setcontext one, just distracted by the concurrent meeting. In that case IP0/IP1 aren't a concern, but lazy binding still is. Specifically, temporary registers aren't saved by _rtld_bind_start, because they're not meant to be live across calls. To avoid that you'd either need to make an indirect call or backport and use variant PCS support (https://reviews.freebsd.org/D44869 has the stack).

Reply:
(Also, re "marking unw_getcontext as a trusted function in the c18n policy, so it should never go through the PLT", it'll still go though a PLT, just not through an additional compartment switch trampoline)

Reply:
> Oh I meant to comment on the setcontext one, just distracted by the concurrent meeting. In that case IP0/IP1 aren't a concern, but lazy binding still is. Specifically, temporary registers aren't saved by _rtld_bind_start, because they're not meant to be live across calls. To avoid that you'd either need to make an indirect call or backport and use variant PCS support (https://reviews.freebsd.org/D44869 has the stack).

Ah. That's slightly annoying. How do we want to approach this for the release? Both of these sound a bit risky w.r.t. regressions. It seems unlikely that exception handling code will care about this in any meaningful way, but it does break libunwind API semantics...